service/controller/orgcluster/cluster.go

```diff
@@ -0,0 +1,72 @@
+package orgcluster
+
+import (
+	// If your operator watches a CRD import it here.
+	// "github.com/giantswarm/apiextensions/v2/pkg/apis/application/v1alpha1"
+
+	"github.com/giantswarm/apiextensions/v3/pkg/label"
+	"github.com/giantswarm/k8sclient/v5/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/operatorkit/v4/pkg/controller"
+	"github.com/giantswarm/operatorkit/v4/pkg/resource"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	cluster "sigs.k8s.io/cluster-api/api/v1alpha3"
+
+	"github.com/giantswarm/rbac-operator/pkg/project"
+)
+
+type OrgClusterConfig struct {
+	K8sClient k8sclient.Interface
+	Logger    micrologger.Logger
+}
+type OrgCluster struct {
+	*controller.Controller
+}
+
+func NewOrgCluster(config OrgClusterConfig) (*OrgCluster, error) {
+	var err error
+
+	var resources []resource.Interface
+	{
+		c := orgClusterResourcesConfig(config)
+
+		resources, err = newOrgClusterResources(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var orgClusterController *controller.Controller
+	{
+		selector := newSelector(func(labels labels.Labels) bool {
+			return labels.Has(label.Organization)
+		})
+
+		c := controller.Config{
+			K8sClient: config.K8sClient,
+			Logger:    config.Logger,
+			NewRuntimeObjectFunc: func() runtime.Object {
+				return new(cluster.Cluster)
```

Comment:
This controller watches cluster resources. I decided to do it that way since it's easier than watching cluster namespaces (they don't have an organization label)

Reply:
Using `gollum` as an example, I do find `giantswarm.io/organization` labels on the namespace resources:

```yaml
apiVersion: v1
kind: Namespace
metadata:
  finalizers:
  - operatorkit.giantswarm.io/rbac-operator-rbac-controller
  labels:
    app: master
    cluster: 5jka7
    customer: giantswarm
    giantswarm.io/cluster: 5jka7
    giantswarm.io/organization: giantswarm
    kubernetes.io/metadata.name: 5jka7
  name: 5jka7
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
```

I think we should rather make sure that we can watch namespaces everywhere instead of watching cluster resources. Reason:

- Watching cluster resources means we require a Cluster API dependency.
- Cluster resources are underlying much more changes.
- Cluster resource versions vary between providers.

I'm sure we will save ourselves a lot of trouble if we can avoid watching clusters.

Reply:
I agree with all three points.
However, at least from what I've seen there seems to be no guarantee for the labels on the cluster namespaces. I have seen some that were `managed-by` chart-operator and some by `cluster-apps-operator`. The one you show does not have a `managed-by` label but it does have the org.
The operators that create the namespace are versioned and different per provider as well.
Ensuring the org-label on these namespaces would help but whichever component we would use for that would then have a cluster-api dependency.